File: lib/python/examples/gpio.py
#!/usr/bin/python3 -B

# Copyright 2023 mjbots Robotic Systems, LLC.  info@mjbots.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""
This example commands a single servo at ID #1 using the default
transport to hold the current position indefinitely, and prints the
state of the servo to the console.
"""

import asyncio
import moteus

async def main():
    # Communicate with controller ID 1 on the default transport.
    c = moteus.Controller()

    # This returns the GPIO digital inputs as an array of bytes, one
    # for each auxiliary port.
    gpio_inputs = await c.read_gpio()

    # Expand these as individal pin values for display purposes.
    def display(aux_num = 1, value = None, count = None):
        print(f"AUX{aux_num}")
        for i in range(count):
            print(f"  Pin {i} - {value & (1 << i)}")

    display(1, gpio_inputs[0], 5)
    print()
    display(2, gpio_inputs[1], 4)


    # And to write the values to gpio pins, the
    # `(make|set)_write_gpio` function can be used.  Here, we'll set
    # all possible GPIO digital outputs to 1.
    await c.set_write_gpio(aux1=0x7f, aux2=0x7f)


    # Finally, GPIO values can be included in general query results.
    qr = moteus.QueryResolution()
    qr.aux1_gpio = moteus.INT8
    qr.aux2_gpio = moteus.INT8
    c2 = moteus.Controller(query_resolution=qr)
    result = await c2.query()

    print()
    print("From moteus.Controller.query()")
    display(1, result.values[moteus.Register.AUX1_GPIO_STATUS], 5)
    print()
    display(2, result.values[moteus.Register.AUX2_GPIO_STATUS], 4)


if __name__ == '__main__':
    asyncio.run(main())

File: lib/rust/examples/gpio.rs
// Copyright 2026 mjbots Robotic Systems, LLC.  info@mjbots.com
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! This example demonstrates reading and writing GPIO digital pins.
//!
//! Usage:
//!   tools/bazel run //lib/rust:gpio
//!   tools/bazel run //lib/rust:gpio -- --fdcanusb /dev/ttyACM0
//!   tools/bazel run //lib/rust:gpio -- --id 1
//!   tools/bazel run //lib/rust:gpio -- --async

use clap::Parser;
use moteus::query::QueryFormat;
use moteus::{AsyncController, BlockingController, Resolution, TransportArgs};

/// Demonstrate GPIO reading and writing.
#[derive(Parser)]
#[command(about = "Demonstrate GPIO reading and writing")]
struct Args {
    /// Servo ID
    #[arg(long, default_value = "1")]
    id: u8,

    /// Use async transport instead of blocking
    #[arg(long)]
    r#async: bool,

    #[command(flatten)]
    transport: TransportArgs,
}

/// Display GPIO pin states for an auxiliary port.
fn display_gpio(aux_num: u8, value: u8, pin_count: u8) {
    println!("AUX{}", aux_num);
    for i in 0..pin_count {
        let state = if value & (1 << i) != 0 { "HIGH" } else { "LOW" };
        println!("  Pin {} - {}", i, state);
    }
}

fn run_blocking(args: &Args) -> Result<(), moteus::Error> {
    let mut c = BlockingController::with_options(args.id, &args.transport.clone().into());

    // Read GPIO digital inputs from both AUX ports. Each value is a
    // byte where bit N represents pin N's state.
    let (aux1, aux2) = c.read_gpio()?;

    // Display the GPIO inputs as individual pin values.
    display_gpio(1, aux1, 5);
    println!();
    display_gpio(2, aux2, 4);

    // Write GPIO digital outputs (set all outputs high).
    c.set_write_gpio(Some(0x7f), Some(0x7f))?;
    println!("\nSet all GPIO outputs to HIGH");

    // GPIO values can also be included in general query results.
    let controller = moteus::Controller::new(args.id).query_format(QueryFormat {
        aux1_gpio: Resolution::Int8,
        aux2_gpio: Resolution::Int8,
        ..QueryFormat::default()
    });
    let mut c2 = BlockingController::with_controller(controller);

    // Query the controller - response includes GPIO values.
    let result = c2.query()?;

    println!("\nFrom BlockingController::query()");
    display_gpio(1, result.aux1_gpio as u8, 5);
    println!();
    display_gpio(2, result.aux2_gpio as u8, 4);

    Ok(())
}

#[tokio::main]
async fn run_async(args: &Args) -> Result<(), moteus::Error> {
    let mut c =
        AsyncController::with_options(args.id, &args.transport.clone().into());

    // Read GPIO digital inputs from both AUX ports. Each value is a
    // byte where bit N represents pin N's state.
    let (aux1, aux2) = c.read_gpio().await?;

    // Display the GPIO inputs as individual pin values.
    display_gpio(1, aux1, 5);
    println!();
    display_gpio(2, aux2, 4);

    // Write GPIO digital outputs (set all outputs high).
    c.set_write_gpio(Some(0x7f), Some(0x7f)).await?;
    println!("\nSet all GPIO outputs to HIGH");

    // GPIO values can also be included in general query results.
    let controller = moteus::Controller::new(args.id).query_format(QueryFormat {
        aux1_gpio: Resolution::Int8,
        aux2_gpio: Resolution::Int8,
        ..QueryFormat::default()
    });
    let mut c2 =
        AsyncController::with_controller(controller);

    // Query the controller - response includes GPIO values.
    let result = c2.query().await?;

    println!("\nFrom AsyncController::query()");
    display_gpio(1, result.aux1_gpio as u8, 5);
    println!();
    display_gpio(2, result.aux2_gpio as u8, 4);

    Ok(())
}

fn main() -> Result<(), moteus::Error> {
    let args = Args::parse();

    if args.r#async {
        run_async(&args)
    } else {
        run_blocking(&args)
    }
}
